core/server/common/src/main/java/alluxio/master/journal/JournalReader.java

```diff
@@ -11,49 +11,36 @@
 
 package alluxio.master.journal;
 
+import alluxio.exception.InvalidJournalEntryException;
+import alluxio.proto.journal.Journal.JournalEntry;
+
+import java.io.Closeable;
 import java.io.IOException;
 
 import javax.annotation.concurrent.NotThreadSafe;
 
 /**
- * This class manages reading from the journal. The reading must occur in two phases:
- *
- * 1. First, the checkpoint must be read.
- *
- * 2. Afterwards, logs are read in order they were created. Only completed logs are read, so the
- * last log currently being written is not read until it is marked as complete.
+ * This class manages reading from the journal.
  */
 @NotThreadSafe
-public interface JournalReader {
-
-  /**
-   * Checks to see if the journal checkpoint has not been updated. If it has been updated since the
-   * creation of this reader, this reader is no longer valid.
-   *
-   * @return true if the checkpoint has not been updated
-   */
-  boolean isValid();
-
+public interface JournalReader extends Closeable {
   /**
-   * Gets the {@link JournalInputStream} for the journal checkpoint. This must be called before
-   * calling {@link #getNextInputStream()}.
+   * Reads an entry from the journal. Return null if there is no more entry left.
    *
-   * @return the {@link JournalInputStream} for the journal checkpoint
-   * @throws IOException if the checkpoint cannot be read, or was already read
+   * @return the journal entry, null if no more entry left
+   * @throws IOException if it failed to read an entry due to an I/O error
+   * @throws InvalidJournalEntryException if the journal entry is invalid (e.g. corrupted entry)
    */
-  JournalInputStream getCheckpointInputStream() throws IOException;
+  JournalEntry read() throws IOException, InvalidJournalEntryException;
 
   /**
-   * @return the input stream for the next completed log. Will return null if the next
-   *         completed log does not exist yet.
-   * @throws IOException if the reader is no longer valid or when trying to get an input stream
-   *                     before a checkpoint was read
+   * @return the next sequence number, can be called after the reader is closed
    */
-  JournalInputStream getNextInputStream() throws IOException;
+  long getNextSequenceNumber();
 
   /**
-   * @return the last modified time of the checkpoint in ms
-   * @throws IOException if the checkpoint does not exist
+   * @return the next sequence number to checkpoint
+   * @throws IOException if an I/O error occur
    */
-  long getCheckpointLastModifiedTimeMs() throws IOException;
+  long getNextSequenceNumberToCheckpoint() throws IOException;
```

Comment:
Semantically, what is the "next sequence number to checkpoint"? Is it any SN after the last existing checkpoint?

Also, this should be determined by whichever client that is deciding to checkpoint or not? As you said, this is not related to this reader.